### PR TITLE
Add UUID and User based economy methods

### PR DESCRIPTION
**Pull request by JRoy**
Adds methods to the Economy API to use UUIDs and User objects. Additionally, deprecates all the username (String) based methods

### CODE REVIEW

**Review by squidicuzz — DISMISSED** [review]
I agree that fetching the economy user on all calls with `getUserByName()` should assure that legacy lookup or non-players are loaded before making any balance changes.  This might resolve intermittent issues with balances corruption if the lookup failed due to not being loaded for some reason.
Looks good.

File: Essentials/src/com/earth2me/essentials/api/Economy.java [hunk 1]
```diff
@@ -86,9 +88,20 @@ private static User getUserByName(String name) {
         return user;
     }
 
+    private static User getUserByUUID(UUID uuid) {
+        if (ess == null) {
+            throw new RuntimeException(noCallBeforeLoad);
+        }
+        if (uuid == null) {
+            throw new RuntimeException("Economy uuid cannot be null");
```
Comment (Ichbinjoe) [inline Essentials/src/com/earth2me/essentials/api/Economy.java]: Don't use RuntimeException for this. IllegalArgumentException is a far better choice.

File: Essentials/src/com/earth2me/essentials/api/Economy.java [hunk 2]
```diff
@@ -133,11 +160,50 @@ public static void setMoney(String name, double balance) throws UserDoesNotExist
...
+     * Sets the balance of a user
+     *
+     * @param user    User
+     * @param balance The balance you want to set
+     *
+     * @throws NoLoanPermittedException If the user is not allowed to have a negative balance
+     */
+    public static void setMoney(User user, BigDecimal balance) throws NoLoanPermittedException {
+        if (user == null) {
+            throw new RuntimeException("Economy user cannot be null");
```
Comment (Ichbinjoe) [inline Essentials/src/com/earth2me/essentials/api/Economy.java]: IllegalArgumentException

File: Essentials/src/com/earth2me/essentials/api/Economy.java [hunk 3]
```diff
@@ -170,17 +238,64 @@ public static void add(String name, double amount) throws UserDoesNotExistExcept
...
+
+    /**
+     * Adds money to the balance of a user
+     *
+     * @deprecated Usernames can change, use {@link Economy#add(UUID, BigDecimal)} or {@link Economy#add(User, BigDecimal)}
+     *
+     * @param user   User
+     * @param amount The money you want to add
+     *
+     * @throws NoLoanPermittedException  If the user is now allowed to have a negative balance
```
Comment (Ichbinjoe) [inline Essentials/src/com/earth2me/essentials/api/Economy.java]: user is *not* allowed

File: Essentials/src/com/earth2me/essentials/api/Economy.java [hunk 4]
```diff
@@ -170,17 +238,64 @@ public static void add(String name, double amount) throws UserDoesNotExistExcept
...
+    /**
+     * Adds money to the balance of a user
+     *
+     * @deprecated Usernames can change, use {@link Economy#add(UUID, BigDecimal)} or {@link Economy#add(User, BigDecimal)}
+     *
+     * @param name   Name of the user
+     * @param amount The money you want to add
+     *
+     * @throws UserDoesNotExistException If a user by that name does not exist
+     * @throws NoLoanPermittedException  If the user is now allowed to have a negative balance
```
Comment (Ichbinjoe) [inline Essentials/src/com/earth2me/essentials/api/Economy.java]: *not*

File: Essentials/src/com/earth2me/essentials/api/Economy.java [hunk 5]
```diff
@@ -170,17 +238,64 @@ public static void add(String name, double amount) throws UserDoesNotExistExcept
         }
     }
 
+    /**
+     * Adds money to the balance of a user
+     *
+     * @deprecated Usernames can change, use {@link Economy#add(UUID, BigDecimal)} or {@link Economy#add(User, BigDecimal)}
+     *
+     * @param name   Name of the user
+     * @param amount The money you want to add
```
Comment (Ichbinjoe) [inline Essentials/src/com/earth2me/essentials/api/Economy.java]: Consider rewording all around to something more like 'The amount of money to be added to the user's account'

**Review by Ichbinjoe — CHANGES_REQUESTED** [review]

**Comment by JRoy** [comment]
Changes applied!

**Review by mdcfe — DISMISSED** [review]

File: Essentials/src/com/earth2me/essentials/api/Economy.java (hunk 4 above)
Comment (mdcfe) [reply Essentials/src/com/earth2me/essentials/api/Economy.java]: @JRoy Is this fixed?

Comment (JRoy) [reply Essentials/src/com/earth2me/essentials/api/Economy.java]: @md678685 it is now :P

**Review by mdcfe — DISMISSED** [review]

**Comment by JRoy** [comment]
@md678685 Backwards compatibility has been maintained and I added User validation to UUID-based methods and null checks in User-based methods

**Review by mdcfe — APPROVED** [review]

**Pull request merged**